match/shreveport_pd.py

```diff
@@ -1,82 +1,78 @@
-        .pipe(gen_uid, ['agency', 'tracking_number', 'allegation'], 'complaint_uid')
-    cprr = match_cprr_post(cprr, post)
-    ensure_data_dir('match')
-    cprr.to_csv(data_file_path(
-        'match/cprr_shreveport_pd_2018_2019.csv'
-    ), index=False)
+import sys
+
+import pandas as pd
+from datamatch import (
+    ThresholdMatcher, JaroWinklerSimilarity, ColumnsIndex, StringSimilarity
+)
+
+from lib.path import data_file_path, ensure_data_dir
+from lib.uid import gen_uid
+
+sys.path.append('../')
+
+
+def match_cprr_codebook(cprr, cb):
+    dfa = cprr[['allegation']]
+    allegation_df = cprr.allegation.str.extract(r'(\d{3}\.\d{1,2})(.+)$')
+    dfa.loc[:, 'name'] = allegation_df.loc[:, 1].str.strip()
+    dfa.loc[:, 'code'] = allegation_df.loc[:, 0]
+    dfa = dfa[dfa.code.notna()].drop_duplicates().set_index('allegation', drop=True)
+
+    matcher = ThresholdMatcher(ColumnsIndex('code'), {
+        'name': StringSimilarity()
+    }, dfa, cb)
+    decision = 0.541
+    lower_bound = 0.5
+    matcher.save_pairs_to_excel(data_file_path(
+        'match/shreveport_pd_cprr_2018_2019_v_codebook.xlsx'
+    ), decision, lower_bound=lower_bound)
+    matches = matcher.get_index_pairs_within_thresholds(decision)
+    match_dict = dict(matches)
+
+    code_n_name = cb.code.str.cat(cb.name, sep=' ')
+    cprr.loc[:, 'allegation'] = cprr.allegation.map(
+        lambda x: x if x not in match_dict else code_n_name[match_dict[x]]
+    )
+    return cprr
+
+
+def match_cprr_post(cprr, post):
+    dfa = cprr[['uid', 'first_name', 'last_name']]
+    dfa = dfa.drop_duplicates().set_index('uid')
+    dfa.loc[:, 'fc'] = dfa.first_name.fillna('').map(lambda x: x[:1])
+
+    dfb = post[['uid', 'first_name', 'last_name']]
+    dfb = dfb.drop_duplicates().set_index('uid')
+    dfb.loc[:, 'fc'] = dfb.first_name.fillna('').map(lambda x: x[:1])
+
+    matcher = ThresholdMatcher(ColumnsIndex('fc'), {
+        'first_name': JaroWinklerSimilarity(),
+        'last_name': JaroWinklerSimilarity()
+    }, dfa, dfb)
+    decision = 0.95
+    matcher.save_pairs_to_excel(data_file_path(
+        'match/shreveport_pd_cprr_2018_2019_v_post_pprr_2020_11_06.xlsx'
+    ), decision)
+    matches = matcher.get_index_pairs_within_thresholds(decision)
+    match_dict = dict(matches)
+
+    cprr.loc[:, 'uid'] = cprr.uid.map(lambda x: match_dict.get(x, x))
+    return cprr
+
+
+if __name__ == '__main__':
+    cprr = pd.read_csv(data_file_path(
+        'clean/cprr_shreveport_pd_2018_2019.csv'
+    ))
+    post = pd.read_csv(data_file_path(
+        'clean/pprr_post_2020_11_06.csv'
+    ))
+    cb = pd.read_csv(data_file_path(
+        'clean/cprr_codebook_shreveport_pd.csv'
+    ))
+    cprr = match_cprr_codebook(cprr, cb)\
+        .pipe(gen_uid, ['agency', 'tracking_number', 'allegation'], 'complaint_uid')
+    cprr = match_cprr_post(cprr, post)
+    ensure_data_dir('match')
+    cprr.to_csv(data_file_path(
+        'match/cprr_shreveport_pd_2018_2019.csv'), index=False)
```

Comment:
Again, this file seems to be plagued with Windows newline sequence.

Reply:
Apparently. No idea why because I had set a default setting. I've gone ahead and standardized the lines formation for the to_csv functions in the match directory, to match your formatting from the match/baton_rouge_pd.py script. What do you think?

Reply:
The line format isn't too important for me as long as they pass flake8 ;)

So you have no idea why this script has Windows's `\r\n`? Anyway, just change the newlines back to `\n` for now.

Reply:
I have an inkling of an idea. 

I seldom receive a message that VS is reverting my lines to` \r\n ` and so I manually revert them back to `\n`. Where my understanding becomes fuzzy is when I try to figure out why it happens, infrequently at that, especially considering that I have changed my default settings. Will continue to figure out why this happens.  

I corrected the newline for `match/shreveport_pd`. Additionally, in the interest of standardization, I reformatted the lines in each match script to resemble the format that you originally used in `baton_rouge_pd`, which you'll see in the commit "standardize to_csv lines in match directory". If this was unnecessary, I'm happy to revert the changes.

Reply:
IMO, we should never spend time manually changing the code styling but rather leave it to auto-formatter like [black](https://dev.to/adamlombard/how-to-use-the-black-python-code-formatter-in-vscode-3lo0). I will open a PR for this so you can take a look.

Reply:
Makes sense. That'd be awesome. Thanks.